src/realm/object-store/sync/impl/emscripten/socket_provider.cpp

```diff
@@ -0,0 +1,219 @@
+    {
+        auto observer = reinterpret_cast<WebSocketObserver*>(user_data);
+        REALM_ASSERT(!event->isText);
+        observer->websocket_binary_message_received(
+            util::Span(reinterpret_cast<char*>(event->data), event->numBytes));
+        return EM_TRUE;
+    }
+
+    static EM_BOOL error_callback(int, const EmscriptenWebSocketErrorEvent*, void* user_data)
+    {
+        auto observer = reinterpret_cast<WebSocketObserver*>(user_data);
+        observer->websocket_error_handler();
+        return EM_TRUE;
+    }
+
+    static EM_BOOL close_callback(int, const EmscriptenWebSocketCloseEvent* event, void* user_data)
+    {
+        auto observer = reinterpret_cast<WebSocketObserver*>(user_data);
+        REALM_ASSERT(event->code >= 1000 && event->code < 5000);
+        auto status = event->code == 1000 ? Status::OK() : Status(ErrorCodes::Error(event->code), event->reason);
+        observer->websocket_closed_handler(event->wasClean, std::move(status));
+        return EM_TRUE;
+    }
+
+    EMSCRIPTEN_WEBSOCKET_T m_socket;
+    std::shared_ptr<LivenessSentinel> m_sentinel;
+    std::unique_ptr<WebSocketObserver> m_observer;
+};
+} // namespace
+
+EmscriptenSocketProvider::EmscriptenSocketProvider()
+    : SyncSocketProvider()
+    , EmscriptenScheduler()
+{
+}
+
+EmscriptenSocketProvider::~EmscriptenSocketProvider() = default;
+
+SyncSocketProvider::SyncTimer EmscriptenSocketProvider::create_timer(std::chrono::milliseconds delay,
+                                                                     FunctionHandler&& handler)
+{
+    util::EmscriptenScheduler& scheduler = *this;
+    return std::make_unique<EmscriptenTimer>(delay, std::move(handler), scheduler);
+}
+
+std::unique_ptr<WebSocketInterface> EmscriptenSocketProvider::connect(std::unique_ptr<WebSocketObserver> observer,
+                                                                      WebSocketEndpoint&& endpoint)
+{
+    // Convert the list of protocols to a string
+    std::ostringstream protocol_list;
+    protocol_list.exceptions(std::ios_base::failbit | std::ios_base::badbit);
+    protocol_list.imbue(std::locale::classic());
+    if (endpoint.protocols.size() > 1)
+        std::copy(endpoint.protocols.begin(), endpoint.protocols.end() - 1,
+                  std::ostream_iterator<std::string>(protocol_list, ","));
+    protocol_list << endpoint.protocols.back();
+    std::string protocols = protocol_list.str();
+    for (size_t i = 0; i < protocols.size(); i++) {
+        if (protocols[i] == '/')
+            protocols[i] = '#';
+    }
+
+    std::string url =
+        util::format("%1://%2:%3%4", endpoint.is_ssl ? "wss" : "ws", endpoint.address, endpoint.port, endpoint.path);
+
+    EmscriptenWebSocketCreateAttributes attr;
+    emscripten_websocket_init_create_attributes(&attr);
+    attr.protocols = protocols.c_str();
+    attr.url = url.c_str();
+    int result = emscripten_websocket_new(&attr);
+    REALM_ASSERT(result > 0);
```

Comment:
This one is special - if the return value is >0 then it's a socket descriptor, if it's <0 then it's `EMSCRIPTEN_RESULT`. And this function only errors out if websockets are not supported, or `attr` is null. I'm not sure what we could do with this information, so I didn't think to record the actual error code.

Reply:
Well we should at least log it and throw an exception or abort, rather than soldiering on as if nothing happened. Note that `REALM_ASSERT` is probably not sufficient for that because they are removed in release builds. It should only be used for things that we believe to be both impossible and within our control, and I don't think this qualifies. 

Although again, I guess this code isn't 100% production ready anyway, so if you want to just file a ticket to revisit error handling of emscripten calls, and leave as-is for this PR, I'm fine with that.

Reply:
This code is 100% not production ready 😅
But in this particular case I'd argue the error conditions are entirely (Realm) programmer error - we either shipped on a browser target so old it doesn't support websockets, or we didn't initialize the function arguments correctly.
Right now release builds for WebAssembly omit any useful debug information so we lose stack traces and exception types and all of that, so even an `REALM_ASSERT_RELEASE` doesn't give us very actionable information. That is something we need to explore more when we tune file size. That's why I'd prefer to have just enough validation for Debug builds for now.

Reply:
> Or do we have checks to ensure that we are using a new enough browser with sufficient support for websockets?

In our particular case the JS SDK itself won't run on a browser that old. But if I think about it some more, this will error out on Node.js as well, since that doesn't come with websockets out of the box.

> Don't we still get logging of the file+line that failed?

If we put the assert inside a convenience function that checks `EMSCRIPTEN_RESULT` wouldn't we only get the line number of that function?